src/nodejs/jshttpclient.nim

```diff
@@ -1,5 +1,212 @@
+when not defined(js):
+  {.fatal: "Module jshttpclient is designed to be used with the JavaScript backend.".}
+
+import std/[asyncjs, jsheaders, jsfetch]
+import fusion/js/jsxmlhttprequest
+from std/uri import Uri
+
+type
+  JsHttpClient* = XMLHttpRequest
+  JsAsyncHttpClient* = ref object of JsRoot
+
+  JsRequest* = ref object of JsRoot
+    url*, `method`*, body*, integrity*, referrer*, mode*, credentials*, cache*, redirect*, referrerPolicy*: cstring
+    headers*: Headers
+    keepalive*: bool
+
+  JsResponse* = ref object of JsRoot
+    status*: cint
+    statusText*, url*, responseText*: cstring
+    headers*: Headers
+
+func newJsHttpClient*(): JsHttpClient {.importjs: "new XMLHttpRequest()".}
+
+func newJsAsyncHttpClient*(): JsAsyncHttpClient = discard
+
+func newJsRequest*(
+  url, `method`: cstring;
+  body, integrity: cstring = "";
+  referrer: cstring = "client";
+  mode: cstring = "cors";
+  credentials: cstring = "include";
+  cache: cstring = "default";
+  redirect: cstring = "follow";
+  referrerPolicy: cstring = "unsafe-url";
+  headers: Headers = newHeaders();
+  keepalive: bool = false
+): JsRequest =
+  result = JsRequest(
+    url: url,
+    `method`: `method`,
+    body: body,
+    integrity: integrity,
+    referrer: referrer,
+    mode: mode,
+    credentials: credentials,
+    cache: cache,
+    redirect: redirect,
+    referrerPolicy: referrerPolicy,
+    headers: headers,
+    keepalive: keepalive
+  )
+
+func fetchOptionsImpl(request: JsRequest): FetchOptions =
+  unsafeNewFetchOptions(
+    metod = request.`method`,
+    body = request.body,
+    mode = request.mode,
+    credentials = request.credentials,
+    cache = request.cache,
+    referrerPolicy = request.referrerPolicy,
+    keepalive = request.keepalive
+  )
+
+func response(response: XMLHttpRequest): JsResponse =
+  ## Converts `XMLHttpRequest` to `JsResponse`
+  new(result)
+  result.status = response.status
+  result.responseText = response.responseText
+
+proc response(response: Response): JsResponse {.async.} =
+  ## Converts `jsfetch` `Response` to `JsResponse`
+  new(result)
+  result.status = response.status
+  result.statusText = response.statusText
+  result.responseText = await text(response)
+  result.url = response.url
+  result.headers = response.headers
+
+proc request*(client: JsHttpClient; request: JsRequest): JsResponse =
+  ## Request proc for sync `XMLHttpRequest` client
+  client.open(metod = request.`method`, url = request.url, async = false)
+  client.send(body = request.body)
+  return client.response()
+
+proc request*(client: JsAsyncHttpClient; request: JsRequest): Future[JsResponse] {.async.} =
+  ## Request proc for async `jsfetch` client
+  return response(await fetch(request.url, fetchOptionsImpl(request)))
+
+proc head*(client: JsHttpClient | JsAsyncHttpClient; url: Uri | string): Future[JsResponse] {.multisync.} =
```

Comment:
It doesn't seem to in this case for some reason, however I have been able to use it for API bindings that compile to JS, is there a reason for this? 
why does it work sometimes but not others ...

Reply:
I think multisync does not work for JS.

You will need to duplicate the all the `func`, is more verbose but it works.

Reply:
I think we should make a multisync macro for this repo, it will be useful.
Here's an example at [L14 - 83](https://gitlab.com/tandy1000/listenbrainz-nim/-/blob/master/src/listenbrainz/utils/client/js.nim#L14).

Reply:
sure